test/jdk/java/lang/Thread/crac/JoinOrSleepOnCRPauseTest.java

```diff
@@ -0,0 +1,186 @@
+
+    private final static long EPS_NS = Long.parseLong(System.getProperty(
+        "test.jdk.java.lang.Thread.crac.JoinOrSleepOnCRPauseTest.eps",
+        "100000000")); // default: 0.1s
+
+    private static final long CRPAUSE_MS = 4000;
+
+    private static final long T_MS = CRPAUSE_MS / 2;
+    private static final  int T_NS = 100;
+
+    private volatile long tDone = -1;
+
+    private final CountDownLatch checkpointLatch = new CountDownLatch(1);
+
+
+    private JoinOrSleepOnCRPauseTest(TestType testType) {
+        this.testType = testType;
+    }
+
+    private void runTest() throws Exception {
+
+        String op = testType.name();
+
+        Thread mainThread = Thread.currentThread();
+
+        Runnable r = () -> {
+
+            try {
+
+                checkpointLatch.countDown();
+
+                switch (testType) {
+
+                    case join_ms:
+                        mainThread.join(T_MS);
+                        break;
+
+                    case join_ns:
+                        mainThread.join(T_MS, T_NS);
+                        break;
+
+                    case sleep_ms:
+                        Thread.sleep(T_MS);
+                        break;
+
+                    case sleep_ns:
+                        Thread.sleep(T_MS, T_NS);
+                        break;
+
+                    default:
+                        throw new IllegalArgumentException("unknown test type");
+                }
+
+            } catch (InterruptedException ie) {
+                throw new RuntimeException(ie);
+            }
+
+            tDone = System.nanoTime();
+        };
+
+        Thread t = new Thread(r);
+        t.start();
+
+        // this additional synchronization is probably redundant;
+        // adding it to ensure we get the expected TIMED_WAITING state
+        // from "our" join or sleep
+        checkpointLatch.await();
+
+        // it is expected that EPS_NS is enough to complete the join/sleep
+        // on restore => expecting that 5 * EPS_NS is enough to enter them
+        long dt = 5 * EPS_NS;
+        Thread.sleep(dt / 1_000_000);
+
+        if (t.getState() != Thread.State.TIMED_WAITING) {
+            throw new RuntimeException("was not able to enter " + op
+                + " in " + dt + " ns");
+        }
+
+        long tBeforeCheckpoint = System.nanoTime();
+
+        jdk.crac.Core.checkpointRestore();
+
+        long tAfterRestore = System.nanoTime();
+
+        t.join();
+
+        long pause = tAfterRestore - tBeforeCheckpoint;
+        if (pause < 1_000_000 * CRPAUSE_MS - EPS_NS) {
+            throw new AssertionError(
+                "the CR pause was less than " + CRPAUSE_MS + " ms");
+        }
+
+        if (tDone < tBeforeCheckpoint + EPS_NS) {
+            throw new AssertionError(
+                op + " has finished before the checkpoint");
+        }
+
+        long eps = Math.abs(tAfterRestore - tDone);
+
+        if (eps > EPS_NS) {
+            throw new AssertionError(
```

Comment:
I think this should be RuntimeError (can happen) vs other AssertionError (cannot)

Reply:
changed to RuntimeException